---
title: "R Notebook"
output: html_notebook
---


```{r}
# Input requirements: =========================================================
# 1. gtf_annot_file: generated during stringtie_merge.sh step, using gffcompare with stringtie --merge output GTF as input and ensembl ref GTF as reference
# 2. gtf_novel_file: generated during stringtie_merge.sh step, using gffcompare with individual sample StringTie GTFs as input and novel annotated GTF (input 1.) as reference
# 3. orig_tracking_file: generated in same way as gtf_annot_file
# 4. combined_tracking_file: generated in same way as gtf_novel_file
# 5. min_occurrence: minimum number of samples in which a transcript must be found to be included in the final GTF
# 6. outfile: output file name
#
# Output: =====================================================================
# 1. outfile: custom annotated GTF

# Load required packages
message(paste(Sys.time(), "Loading required libraries ..."), sep = "\t")
suppressPackageStartupMessages({
  library(tidyverse)
  library(GenomicRanges)
  library(rtracklayer)
  library(data.table)
})
```


```{r}
# Global arguments
args <- commandArgs(trailingOnly = TRUE)

# functions_file = args[1]
functions_file = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/scripts/02_rnaseq_assembly/filter_annotate_functions.R"

# merged_basename = args[2]
merged_basename = "NBL_transcriptome_full_new"

# gtf_ref_loc = args[3]
gtf_ref_loc = "/hpc/pmc_vanheesch/shared_resources/GENOMES/Homo_sapiens.GRCh38/102/annotation/Homo_sapiens.GRCh38.102.gtf"

# gtf_refseq_basename = args[4]
gtf_refseq_basename = "/hpc/pmc_vanheesch/shared_resources/GENOMES/Homo_sapiens.GRCh38/102/annotation/Homo_sapiens.GRCh38.p13"

# gtf_annot_file = args[5]
gtf_annot_file = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_transcriptome_full/NBL_transcriptome_full_new.annotated.gtf"

# gtf_novel_file = args[6]
gtf_novel_file = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_transcriptome_full/NBL_transcriptome_full_new_combined.combined.gtf"

# orig_tracking_file = args[7]
orig_tracking_file = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_transcriptome_full/NBL_transcriptome_full_new.tracking"

# combined_tracking_file = args[8]
combined_tracking_file = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_transcriptome_full/NBL_transcriptome_full_new_combined.tracking"

# min_occurrence = args[9]
min_occurrence = 4

outfile = args[10]

min_occurrence <- as.numeric(min_occurrence)
```


```{r}
# Load filtering functions
source(functions_file)


# Load files
message(paste(Sys.time(), "Loading files ..."), sep = "\t")

## Load annotated novel GTF file
gtf_annot <- rtracklayer::import(gtf_annot_file)
gtf_annot <- data.table::as.data.table(gtf_annot)
```


```{r}
## Load combined GTF file (generated with gffcompare, using individual sample StringTie GTFs as input and novel annotated GTF as reference)
gtf_novel_GR <- rtracklayer::import(gtf_novel_file)
```


```{r}
## Load original tracking file
orig_tracking <- data.table::fread(orig_tracking_file, header = F, select = c(1:5)) %>%
  tidyr::separate(V3, into = c("ref_gene_id_annotated", "ref_transcript_id_annotated"), sep = "\\|") %>%
  tidyr::separate(V5, into = c("query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len"), sep = "\\|")
colnames(orig_tracking) <- c("TCONS", "xloc", "ref_gene_id_annotated", "ref_transcript_id_annotated", "class_code_annotated", "query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len")
orig_tracking$query_gene_id <- sub("^q1:", "", orig_tracking$query_gene_id)

### Merge orig_tracking_file with annotated GTF (generated with gffcompare using stringtie --merge GTF as input and ref GTF as reference)
orig_tracking_annot <- orig_tracking %>%
  left_join(subset(gtf_annot, type == "transcript")[, c("gene_name", "ref_gene_id", "cmp_ref", "cmp_ref_gene", "transcript_id")], by = c("query_transcript_id" = "transcript_id"))

### Remove redundant lines from orig_tracking_annot
# orig_tracking_annot <- orig_tracking_annot %>%
#   group_by(TCONS) %>%
#   mutate(
#     gene_name = first(gene_name[gene_name != "NA"]),
#     ref_gene_id = first(ref_gene_id[ref_gene_id != "NA"]),
#     cmp_ref = first(cmp_ref[cmp_ref != "NA"]),
#     cmp_ref_gene = first(cmp_ref_gene[cmp_ref_gene != "NA"]),
#   ) %>%
#   ungroup() %>%
#   distinct()



```


```{r}
## Load combined gffcompare tracking file
combined_tracking <- data.table::fread(combined_tracking_file, header = F, select = c(1:5)) %>%
  tidyr::separate(V3, into = c("combined_gene_id", "ref_transcript_id"), sep = "\\|") %>%
  tidyr::separate(V5, into = c("query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len"), sep = "\\|")
colnames(combined_tracking) <- c("TCONS_combined", "xloc", "combined_gene_id", "combined_transcript_id", "class_code_combined", "query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len")
```


```{r}
## Merge combined_tracking file with orig_tracking file to annotate combined GTF
combined_tracking_annotated <- combined_tracking[, c("TCONS_combined", "combined_gene_id", "combined_transcript_id", "class_code_combined")] %>%
  left_join(orig_tracking_annot[, c("ref_gene_id_annotated", "ref_transcript_id_annotated", "query_gene_id", "query_transcript_id", "class_code_annotated", "gene_name", "cmp_ref", "cmp_ref_gene")], na_matches = "never", by = c("combined_transcript_id" = "query_transcript_id"))
```


```{r}
## Load reference GTF
gtf_reference <- rtracklayer::import.gff("/hpc/pmc_vanheesch/shared_resources/GENOMES/Homo_sapiens.GRCh38/102/annotation/Homo_sapiens.GRCh38.102.gtf", colnames = c(
  "type",
  "source",
  "gene_id",
  "gene_name",
  "gene_biotype",
  "transcript_id",
  "transcript_name"
))
gtf_ref_df <- as.data.frame(gtf_reference)
```


```{r}
# Filtering and annotation #####################################################
message(paste(Sys.time(), "Removing transcripts on scaffolds ..."), sep = "\t")
gtf_novel_GR <- gtf_novel_GR[seqnames(gtf_novel_GR) %in% c(1:22, "X", "Y"), ]

message(paste(Sys.time(), "Removing unstranded transcripts ..."), sep = "\t")
gtf_novel_GR <- gtf_novel_GR[strand(gtf_novel_GR) != "*", ]

gtf_novel <- data.table::as.data.table(gtf_novel_GR)
gtf_novel$num_samples <- as.numeric(gtf_novel$num_samples)
```


```{r}
# Remove low abundance transcripts
message(paste0(Sys.time(), "\t", "Filtering transcripts found in fewer than ", min_occurrence, " samples ..."))
transcripts_keep <- subset(gtf_novel, type == "transcript" & (num_samples >= min_occurrence | class_code == "="))$transcript_id
gtf_novel <- gtf_novel[which(gtf_novel$transcript_id %in% transcripts_keep),]
```


```{r}
## Set exon cmp_ref IDs and transcript_ids same as parent transcript
gtf_novel <- gtf_novel %>%
  group_by(transcript_id) %>%
  mutate(
    gene_name = first(gene_name[gene_name != "NA"]),
    cmp_ref = first(cmp_ref[cmp_ref != "NA"]),
    cmp_ref_gene = first(cmp_ref_gene[cmp_ref_gene != "NA"]),
    # oId = first(oId[oId != "NA"])
  ) %>%
  ungroup()
```


```{r}
## Merge gtf_novel with combined tracking files
gtf_novel <- gtf_novel %>%
  dplyr::left_join(combined_tracking_annotated, suffix = c(".gtf", ".annot"), by = c("transcript_id" = "TCONS_combined"))
```

```{r}
## Remove unwanted transcript classes
transcripts_discard <- gtf_novel[which(gtf_novel$type == "transcript" &
                                                 gtf_novel$class_code_annotated %in% c("=", "c", "j", "m", "n", "e", "r", "s")), ]
gtf_novel <- gtf_novel[which(!(gtf_novel$transcript_id %in% transcripts_discard$transcript_id)), ]

## Remove transcripts annotated as 'x-class STRG of x-class MSTRG'
TCONS_discard <- subset(combined_tracking_annotated, class_code_combined == "x" & class_code_annotated == "x")$TCONS_combined

gtf_novel <- gtf_novel[which(!(gtf_novel$transcript_id %in% TCONS_discard)), ]
```


```{r}
# Convert data frames to data tables
setDT(gtf_novel)
setDT(gtf_ref_df)

# Filter out rows with NA values in ref_gene_id_annotated
gtf_novel <- gtf_novel[!is.na(ref_gene_id_annotated)]

# Rename strand in gtf_novel to avoid column name conflict
setnames(gtf_novel, "strand", "strand_novel")

# Create a unique reference table for gene_id and strand
unique_gtf_reference <- unique(gtf_ref_df[, .(gene_id, strand)])

# Join with gtf_novel using a binary search for speed
setkey(gtf_novel, ref_gene_id_annotated)
setkey(unique_gtf_reference, gene_id)
gtf_novel <- gtf_novel[unique_gtf_reference, nomatch=0]

# Update the gene_id column using vectorized operations
gtf_novel[, gene_id := ifelse(
  ref_gene_id_annotated != "-" & strand_novel == strand, 
  ref_gene_id_annotated, 
  combined_gene_id
)]
```

```{r}
# Alternative
# Reduce gtf_reference to unique gene_id - strand pairs
gtf_reference_unique <- gtf_ref_df %>%
  select(gene_id, strand) %>%
  distinct()

# Join the gtf_novel with the unique gtf_reference on ref_gene_id_annotated
gtf_novel_joined <- left_join(gtf_novel, gtf_reference_unique, by = c("ref_gene_id_annotated" = "gene_id"))

gtf_novel_test <- gtf_novel_joined %>%
  group_by(transcript_id) %>%
  filter(!is.na(ref_gene_id_annotated)) %>%
  # Remove x and i transcripts that have same strand as reference
  filter(
    is.na(strand.y) |
    (class_code_annotated %in% c('x', 'i') & strand.x != strand.y) |
    (!class_code_annotated %in% c('x', 'i') & strand.x == strand.y)
  ) %>%
  mutate(
    # Determine the gene_id based on conditions
    gene_id = case_when(
      is.na(strand.y) ~ combined_gene_id,
      class_code_annotated %in% c('x', 'i') & strand.x != strand.y ~ combined_gene_id,
      !class_code_annotated %in% c('x', 'i') & strand.x == strand.y & ref_gene_id_annotated != "-" ~ ref_gene_id_annotated,
      TRUE ~ combined_gene_id
    )
  )

# %>%
  # Drop the extra columns from the join
  # select(-strand.y)
```

```{r}
# Set unique transcript IDs for each TCONS ID (gffcompare finds different TCONSs within same annotated parent transcript isoforms, resulting in multiple different isoforms with same MSTRG.X.X ID if not fixed)
# gtf_novel_test <- gtf_novel_test

setDT(gtf_novel_test)
setorder(gtf_novel_test, combined_gene_id, transcript_id)
gtf_novel_test[, counter := cumsum(!duplicated(transcript_id)), by = combined_gene_id]
gtf_novel_test[, transcript_id_new := paste0(combined_gene_id, ".", counter)]
gtf_novel_test[, counter := NULL]
```


```{r}
## Annotate gene IDs and transcript IDs
gtf_novel_test <- gtf_novel_test %>%
  group_by(transcript_id) %>%
  mutate(
    ref_gene_id = ref_gene_id_annotated,
    # gene_id = ifelse(!is.na(ref_gene_id), ref_gene_id, combined_gene_id),
    ref_ranscript_id = ref_transcript_id_annotated,
    transcript_id = transcript_id_new,
    gene_name = ifelse(!is.na(gene_name.annot), 
                       gene_name.annot,
                       ifelse(!is.na(gene_name.gtf), gene_name.gtf, cmp_ref_gene.gtf)),
    class_code = first(class_code[class_code != "NA"]),
    class_code_annotated = first(class_code_annotated[class_code_annotated != "NA"]),
    # gene_name = ifelse(is.na(gene_name), gene_id, gene_name)
    ) %>%
  ungroup()
```


```{r}
## Select relevant columns
gtf_novel_test <- gtf_novel_test %>%
  mutate(class_code = class_code_annotated) %>%
  dplyr::select(c(seqnames, start, end, width, strand.x, source, type, score, phase, gene_id, gene_name, transcript_id, ref_gene_id, class_code, tss_id, num_samples, exon_number, cmp_ref.annot, cmp_ref_gene.annot))
colnames(gtf_novel_test) <- c("seqnames", "start", "end", "width", "strand", "source", "type", "score", "phase", "gene_id", "gene_name", "transcript_id", "ref_gene_id", "class_code", "tss_id", "num_samples", "exon_number", "cmp_ref", "cmp_ref_gene")
```


```{r}
## Flagging RefSeq transcripts
message(paste(Sys.time(), "Flagging XR transcript overlap ..."), sep = "\t")
gtf_novel_test <- annotate_overlap(gtf = gtf_novel_test,
                              gtf_refseq_basename = gtf_refseq_basename,
                              x_name = "xr")

message(paste(Sys.time(), "Flagging NR transcript overlap ..."), sep = "\t")
gtf_novel_test <- annotate_overlap(gtf = gtf_novel_test,
                              gtf_refseq_basename = gtf_refseq_basename,
                              x_name = "nr")

gtf_novel_test <- data.table::as.data.table(gtf_novel_test)
```


```{r}
## Add biotype to custom annotation based on reference ID 
message(paste(Sys.time(), "Adding biotype to StringTie transcripts ..."),
        sep = "\t")
gtf_novel_test$gene_biotype <-
  gtf_reference$gene_biotype[match(gtf_novel_test$gene_id,
                                   gtf_reference$gene_id)]
gtf_novel_test[which(is.na(gtf_novel_test$gene_biotype)), "gene_biotype"] <- "stringtie"
```


```{r}
## Check ref tx overlap 
message(paste(Sys.time(), "Checking reference transcript overlap ..."),
        sep = "\t")
ref_overlap_txs <- suppressWarnings(check_tx_overlap(gtf = gtf_novel_test,
                                    gtf_reference = gtf_reference))
```
```{r}
gtf_novel_test <- subset(
  gtf_novel_test, !(gtf_novel_test$transcript_id %in% ref_overlap_txs)
)
```

```{r}
## Remove genes with exons on both strands
# 1. Subset the GTF data to retain only exon entries
exon_data <- gtf_novel_test[gtf_novel_test$type == "exon",]

# 2. Group by gene and reference sequence, 
# 3. Count unique strands for each gene-reference sequence combination
gene_strand_count <- exon_data %>%
  group_by(gene_id = exon_data$gene_id, seqnames) %>%
  summarise(n_strands = n_distinct(strand)) %>%
  ungroup()

# 4. Identify genes with exons on both strands
genes_both_strands <- gene_strand_count[gene_strand_count$n_strands > 1,]$gene_id

# 5. Remove genes with exons on both strands
gtf_novel_test <- subset(
  gtf_novel_test, !(gtf_novel_test$gene_id %in% genes_both_strands)
)
```

```{r}
gtf_novel_GR_out <-
  GenomicRanges::makeGRangesFromDataFrame(gtf_novel_test, keep.extra.columns = T)
```


```{r}
# Merge novel GTF with hg38 GTF
gtf_novel_merged <- c(gtf_reference, gtf_novel_GR_out)
```


```{r}
# Export GTF
message(paste(Sys.time(), "Exporting custom gtf ... "), sep = "\t")
rtracklayer::export.gff(object = gtf_novel_merged, con = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/customannotation/NBL_transcriptome_full_novel_filtered_new_withmergedparent.gtf")
message(paste(Sys.time(), "Exporting custom gtf ... Done!"), sep = "\t")

```

