---
title: "R Notebook"
output: html_notebook
---

# use gffcompare combined.gtf instead of tracking_file to look at sample occurrence? 
# or check overlap between the combined.gtf and the sample matrix manually created (use TPM = 0?) to see if they give same values
# would be nice for each transcript (isoform) to have a list of samples it is found in

# load tracking file
```{r}
tracking_file <- read.table(file = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome_combined.tracking", header = F)
```

```{r}
colnames(tracking_file) <-
  c("transfrag_id", "locus_id", "ref_gene_id", "class_code")
# tracking_file <- subset(tracking_file, class_code == "=")
```

# load gffcompare combined gtf
```{r}
combined_gtf <- rtracklayer::import("/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome_vsensembl.combined.gtf")
```

```{r}
combined_gtf_df <- as.data.frame(combined_gtf)

transcripts_keep <- subset(combined_gtf_df, num_samples > 1)$transcript_id

combined_gtf_filtered <- subset(combined_gtf, transcript_id %in% transcripts_keep)

rtracklayer::export.gff(object = combined_gtf_filtered, con = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome_vsensembl.combined.filtered.gtf")

```

```{r}
# Turn tracking_file into data.frame
tracking_file_df <- data.frame(tracking_file, check.names = F)

# Filter tracking_file for transcript_id values that occur in combined_gtf_df_filtered
tracking_file_filtered <- subset(tracking_file_df, transfrag_id %in% combined_gtf_df_filtered$transcript_id)

# Split third column into two columns, using | as delimiter
tracking_file_filtered <- tracking_file_filtered %>%
  separate(V3, into = c("ref_gene_id", "ref_transcript_id"), sep = "\\|")

```

```{r}
tracking_genes <- data.frame(samples = tracking_file[, 1:4])
tracking_genes$rowid <- 1:nrow(tracking_genes)

tracking_samples <-
  data.frame(samples = tracking_file[, 5:length(tracking_file)])
# tracking_samples$rowid <- 1:nrow(tracking_samples)
```


```{r}
tracking_samples_tpm <- sapply(tracking_samples,
                               function(x)
                                 sapply(strsplit(as.character(x), split = "[|]"), "[", 5))

tracking_samples_tpm <- stringr::str_extract(tracking_samples,  "(?<=\\|)[^|]+$")

tracking_samples_tpm <- data.frame(tracking_samples_tpm)
tracking_samples_tpm <-
  dplyr::mutate_all(tracking_samples_tpm, function(x)
    as.numeric(as.character(x)))
```


```{r}
# Count transfrag occurrence
occurrence_df <-
  data.frame(ifelse(tracking_samples_tpm > min_tpm, 1, 0))
occurrence_df[is.na(occurrence_df)] <- 0
occurrence_df <- data.frame(rowSums(occurrence_df))
colnames(occurrence_df) <- "occurrence"
sum(occurrence_df > 2)

# Subset transcripts on their occurence
tracking_genes <- cbind(tracking_genes, occurrence_df)
tracking_genes <-
  tracking_genes[tracking_genes$occurrence >= min_occurence, ]
tracking_genes$transcript_id <-
  gsub(".*\\|", "", tracking_genes$samples.ref_gene_id)

# Grab colnames for sample matrix
sample_ids <- readLines(paste(wd, "documentation/sample_ids.txt", sep = "/"))

# Subset sample matrix and write to tracking file directory
tracking_file <-
  tracking_file[tracking_file$ref_gene_id %in% tracking_genes$samples.ref_gene_id, ]

colnames(tracking_file) <-
  c("transfrag_id",
    "locus_id",
    "ref_gene_id",
    "class_code",
    sample_ids)
```









# extract transcript tpms from tracking file by transcript ID
# make table showing number of samples with TPM > 1 per transcript ID as well as names of samples with TPM > 1





# extract transcript num_samples from combined gtf by transcript ID (or by xloc / transfrag ID? need to check)
# add num_samples to the table made above to compare 