---
title: "R Notebook"
output: html_notebook
---

This notebook contains work in progress pipeline for filtering a GTF file after running stringtie (individual samples) --> gffcompare to ref GTF --> gffcompare individual GTFs to custom annotated GTF

```{r}
library(rtracklayer)
library(dplyr)
library(tidyr)
library(data.table)
library(tidyverse)

args <- commandArgs(trailingOnly = TRUE)

workdir <- "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/"
# workdir = args[1]

merged_basename <- "NBL_full_transcriptome"
# merged_basename = args[2]

setwd(workdir)
```


```{r}
# Load files
# gffcompare_annotated_gtf_loc <- paste0("analysis/gffcompare/", merged_basename, "/", merged_basename, ".annotated.gtf")
gffcompare_annotated_gtf_loc <- "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome_combined.combined.gtf"

# gffcompare_tracking_loc <- paste0("analysis/gffcompare/", merged_basename, "/", merged_basename, ".tracking")
combined_tracking_loc <- "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome_combined.tracking"

orig_tracking_loc <- "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome.tracking"
```


```{r}
# Process GTF file
gffcompare_annotated_gtf <- rtracklayer::import(gffcompare_annotated_gtf_loc)
gffcompare_annotated_gtf_df <- data.table::as.data.table(gffcompare_annotated_gtf)
gffcompare_annotated_gtf_df$num_samples <- as.numeric(gffcompare_annotated_gtf_df$num_samples)
```


```{r}
# Remove transcripts not found in > 2 samples
transcripts_keep <- subset(gffcompare_annotated_gtf_df, type == "transcript" & num_samples > 3)$transcript_id
gtf_keep <- gffcompare_annotated_gtf_df[which(gffcompare_annotated_gtf_df$transcript_id %in% transcripts_keep),] 
```


```{r}
# Set exon cmp_ref IDs and transcript_ids same as parent transcript
gtf_keep_txs <- gtf_keep %>%
  group_by(transcript_id) %>%
  mutate(
    gene_name = first(gene_name[gene_name != "NA"]),
    cmp_ref = first(cmp_ref[cmp_ref != "NA"]),
    cmp_ref_gene = first(cmp_ref_gene[cmp_ref_gene != "NA"]),
    oId = first(oId[oId != "NA"])
  ) %>%
  ungroup()
```


```{r}
annotated_gtf_df <- as.data.table(rtracklayer::import("/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome.annotated.gtf"))
```


```{r}
# Load combined gffcompare tracking file
combined_tracking <- readr::read_delim(combined_tracking_loc, col_names = F, col_select = c(1:5)) %>%
  tidyr::separate(X3, into = c("combined_gene_id", "ref_transcript_id"), sep = "\\|") %>%
  tidyr::separate(X5, into = c("query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len"), sep = "\\|")

colnames(combined_tracking) <- c("TCONS", "xloc", "combined_gene_id", "combined_transcript_id", "class_code_combined", "query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len")
```

```{r}
# Load original tracking file
orig_tracking <- readr::read_delim(orig_tracking_loc, col_names = F, col_select = c(1:5)) %>%
  tidyr::separate(X3, into = c("ref_gene_id_tracking", "ref_transcript_id"), sep = "\\|") %>%
  tidyr::separate(X5, into = c("query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len"), sep = "\\|")

# Set tracking file column names
colnames(orig_tracking) <- c("TCONS", "xloc", "ref_gene_id_orig", "ref_transcript_id", "class_code_orig", "query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len")

orig_tracking$query_gene_id <- sub("^q1:", "", orig_tracking$query_gene_id)

orig_tracking_annot <- orig_tracking %>%
  left_join(annotated_gtf_df[, c("gene_name", "ref_gene_id", "cmp_ref", "cmp_ref_gene", "transcript_id")], by = c("query_transcript_id" = "transcript_id"))

# orig_tracking_annot <- distinct(orig_tracking_annot)

orig_tracking_annot <- orig_tracking_annot %>%
  group_by(TCONS) %>%
  mutate(
    gene_name = first(gene_name[gene_name != "NA"]),
    ref_gene_id = first(ref_gene_id[ref_gene_id != "NA"]),
    cmp_ref = first(cmp_ref[cmp_ref != "NA"]),
    cmp_ref_gene = first(cmp_ref_gene[cmp_ref_gene != "NA"]),
  ) %>%
  ungroup() %>%
  distinct()
```

```{r}
# condition <- !is.na(orig_tracking_annot$ref_gene_id_orig) & 
#              !is.na(orig_tracking_annot$ref_gene_id) & 
#              orig_tracking_annot$ref_gene_id_orig != orig_tracking_annot$ref_gene_id
# 
# # Filter data to see only non-matching rows, excluding NAs
# non_matching_rows <- orig_tracking_annot[condition, ]
```


```{r}
combined_tracking_annotated <- combined_tracking[, c("TCONS", "combined_gene_id", "combined_transcript_id", "class_code_combined")] %>%
  left_join(orig_tracking_annot[, c("ref_gene_id", "ref_transcript_id", "query_gene_id", "query_transcript_id", "class_code_orig", "gene_name", "cmp_ref", "cmp_ref_gene")],na_matches = "never", by = c("combined_transcript_id" = "query_transcript_id"))

# The combined and orig tracking files might not need to be both loaded, I think the combined tracking file in itself should suffice
```



```{r}
gtf_keep_txs <- gtf_keep_txs %>%
  dplyr::left_join(combined_tracking_annotated, suffix = c(".gtf", ".annot"), by = c("transcript_id" = "TCONS"))
```

```{r}
# Remove unwanted transcript classes
transcripts_discard <- gtf_keep_txs[which(gtf_keep_txs$type == "transcript" &
                                                 gtf_keep_txs$class_code_orig %in% c("=", "c", "j", "m", "n", "e", "r", "s")
  ), ]
gtf_keep_txs <-
  gtf_keep_txs[which(!(
    gtf_keep_txs$transcript_id %in% transcripts_discard$transcript_id
  )), ]
```

```{r}
# Annotate gene IDs and transcript IDs
gtf_keep_annotated <- gtf_keep_txs %>%
  group_by(transcript_id) %>%
  mutate(
    ref_gene_id = ref_gene_id,
    gene_id = ifelse(!is.na(ref_gene_id), ref_gene_id, combined_gene_id),
    ref_ranscript_id = ref_transcript_id,
    transcript_id = combined_transcript_id,
    gene_name = ifelse(!is.na(gene_name.annot), gene_name.annot, gene_name.gtf),
    class_code = first(class_code[class_code != "NA"]),
    class_code_orig = first(class_code_orig[class_code_orig != "NA"]),
    # gene_name = ifelse(is.na(gene_name), gene_id, gene_name)
    ) %>%
  ungroup()

gtf_keep_annotated <- subset(gtf_keep_annotated, class_code != "u")  # Need to find out what to do with these guys, as some of them occur in quite a lot of samples
```


```{r}
gtf_keep_final <- gtf_keep_annotated %>%
  mutate(class_code = class_code_orig) %>%
  dplyr::select(c(seqnames, start, end, width, strand, source, type, score, phase, gene_id, gene_name, transcript_id, ref_gene_id, class_code, tss_id, num_samples, exon_number, cmp_ref.annot, cmp_ref_gene.annot))

gtf_keep_final_GR <- GenomicRanges::makeGRangesFromDataFrame(gtf_keep_final, keep.extra.columns = TRUE)

# I like the final GTF, as novel transcripts have unique MSTRG gene IDs and transctipt IDs, but the gene names are set to the gene they overlap if applicable.
# Now just need to integrate further with the rest of the pipeline


rtracklayer::export.gff(object = gtf_keep_final_GR, con = "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/customannotation/NBL_transcriptome_test_4.gtf")

```


