---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
library(rtracklayer)
library(tidyverse)
```


```{r}
GTF_novel <- rtracklayer::import("/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/customannotation/NBL_transcriptome_full_novel_filtered_new.gtf")

merged_novel_df <- as.data.frame(gtf_novel_merged)
```

```{r}
gtf_novel_combined <- rtracklayer::import("/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/analysis/gffcompare/NBL_transcriptome_full/NBL_transcriptome_full_combined.combined.gtf")

gtf_novel_combined_df <- as.data.table(gtf_novel_combined)
```


```{r}
gtf_novel_merged_df <- data.frame(gtf_novel_merged)
# 2. Subset the GTF data to retain only exon entries
exon_data <- gtf_novel_merged_df[gtf_novel_merged_df$type == "exon",]
```


```{r}
# 3. Group by gene and reference sequence, 
# 4. Count unique strands for each gene-reference sequence combination
gene_strand_count <- exon_data %>%
  group_by(gene_id = exon_data$gene_id, seqnames) %>%
  summarise(n_strands = n_distinct(strand)) %>%
  ungroup()
```


```{r}
# 5. Identify genes with exons on both strands
genes_both_strands <- gene_strand_count[gene_strand_count$n_strands > 1,]$gene_id
```

