---
title: "R Notebook"
output: html_notebook
---
This notebook contains work in progress pipeline for filtering a GTF file after running stringtie (individual samples) --> gffcompare individual GTFs to ref GTF (skipping stringtie merge)

```{r}
library(rtracklayer)
library(dplyr)
library(tidyr)
library(data.table)

args <- commandArgs(trailingOnly = TRUE)

workdir <- "/hpc/pmc_vanheesch/projects/Damon/Neuroblastoma_neoantigens/RNA_nbl_transcriptome_assembly/"
# workdir = args[1]

merged_basename <- "NBL_full_transcriptome"
# merged_basename = args[2]

setwd(workdir)
```


```{r}
# Load files
gffcompare_annotated_gtf_loc <- paste0("analysis/gffcompare/", merged_basename, "/", merged_basename, ".annotated.gtf")
gffcompare_annotated_gtf_loc <- "analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome_vsensembl.combined.gtf"

gffcompare_tracking_loc <- paste0("analysis/gffcompare/", merged_basename, "/", merged_basename, ".tracking")
gffcompare_tracking_loc <- "analysis/gffcompare/NBL_full_transcriptome/NBL_full_transcriptome_vsensembl.tracking"
```


```{r}
# Process GTF file
gffcompare_annotated_gtf <- rtracklayer::import(gffcompare_annotated_gtf_loc)
gffcompare_annotated_gtf_df <- data.table::as.data.table(gffcompare_annotated_gtf)
gffcompare_annotated_gtf_df$num_samples <- as.numeric(gffcompare_annotated_gtf_df$num_samples)

# Remove transcripts not found in > 2 samples
transcripts_keep <- subset(gffcompare_annotated_gtf_df, type == "transcript" & num_samples > 3)$transcript_id
gtf_keep <- gffcompare_annotated_gtf_df[which(gffcompare_annotated_gtf_df$transcript_id %in% transcripts_keep),] 

# Set exon cmp_ref IDs and transcript_ids same as parent transcript
gtf_keep_txs <- gtf_keep %>%
  group_by(transcript_id) %>%
  mutate(
    gene_name = first(gene_name[gene_name != "NA"]),
    cmp_ref = first(cmp_ref[cmp_ref != "NA"]),
    cmp_ref_gene = first(cmp_ref_gene[cmp_ref_gene != "NA"]),
    oId = first(oId[oId != "NA"])
  ) %>%
  ungroup()
```


```{r}
# Load gffcompare tracking file
gffcompare_tracking <- readr::read_delim(gffcompare_tracking_loc, col_names = F, col_select = c(1:5)) %>%
  tidyr::separate(X3, into = c("ref_gene_id_tracking", "ref_transcript_id"), sep = "\\|") %>%
  tidyr::separate(X5, into = c("query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len"), sep = "\\|")

# Set tracking file column names
colnames(gffcompare_tracking) <- c("TCONS", "xloc", "ref_gene_id_tracking", "ref_transcript_id", "class_code_orig", "query_gene_id", "query_transcript_id", "query_num_exons", "query_FPKM", "query_TPM", "query_cov", "query_len")
```


```{r}
# Add tracking file annotations to GTF by matching cmp_ref in GTF to query_transcript_id in tracking file
gtf_keep_txs <- gtf_keep_txs %>%
  dplyr::left_join(gffcompare_tracking[, c("TCONS", "ref_gene_id_tracking", "ref_transcript_id", "class_code_orig", "xloc")], by = c("transcript_id" = "TCONS"))
```


```{r}
# Remove unwanted transcript classes
transcripts_discard <- gtf_keep_txs[which(gtf_keep_txs$type == "transcript" &
                                                 gtf_keep_txs$class_code %in% c("=", "c", "j", "m", "n", "e", "r", "s")
  ), ]
gtf_keep_txs <-
  gtf_keep_txs[which(!(
    gtf_keep_txs$transcript_id %in% transcripts_discard$transcript_id
  )), ]
```


```{r}
# Annotate gene IDs and transcript IDs
gtf_keep_annotated <- gtf_keep_txs %>%
  group_by(transcript_id) %>%
  mutate(
    gene_id = ifelse(ref_gene_id_tracking != "-", ref_gene_id_tracking, gsub("\\.[^.]*$", "", oId)),
    transcript_id = ifelse(!is.na(cmp_ref), cmp_ref, oId),
    gene_name = ifelse(!is.na(cmp_ref_gene), cmp_ref_gene, gene_name),
    gene_name = ifelse(is.na(gene_name), gene_id, gene_name)
    ) %>%
  ungroup()
```


```{r}
gtf_keep_final <- gtf_keep_annotated %>%
  dplyr::select(-c(ref_gene_id_tracking, ref_transcript_id, class_code_orig))

gtf_keep_final_GR <- GenomicRanges::makeGRangesFromDataFrame(gtf_keep_final, keep.extra.columns = TRUE)

rtracklayer::export.gff(object = gtf_keep_final_GR, con = "analysis/customannotation/NBL_transcriptome_test.gtf")

```

